README.md
# radioline_map
*radioline_map* reads network topology from master RF-module *RAD-868-IFS* (c) by Phoenix Contact
 combines it with GeoJSON metadata and generates an interactive layer on OSM (Open Street Map).

Copyright (C) 2019 Stefan Braun


## about radioline_map
An INOFFICIAL OpenSource command line application written in Python 3.  

This tool needs a serial connection to master RF-module via *RAD-CABLE-USB* (c) or *IFS-RS232-DATACABLE* (c) by Phoenix Contact.   
It is optimized for usage in our company and works well with *RAD-868-IFS*, it's possible that it works for all industrial radio frequency modules in their "Trusted Wireless 2.0" (c) product line with IFS-dataport.       
There's no warranty that this tool works in every situation!   
   
   
### usage   
```
usage: generate_map.py [-h] [--zoom ZOOM] [--open_browser] [-v] --port COMPORT
                       --mapfile MAPFILE --htmlfile HTMLFILE

Show network topology of RF-modules *RAD-868-IFS* on OSM(Open Street Map).

optional arguments:
  -h, --help            show this help message and exit
  --zoom ZOOM, -z ZOOM  initial zoom factor on map (default: 15)
  --open_browser, -b    open generated map in webbrowser (default: False)
  -v, --verbosity       output verbosity, setting this flag more times means
                        more verbosity (default: loglevel WARNING)

required named arguments:
  --port COMPORT, -p COMPORT
                        serial COM-port connected to IFS-dataport (e.g.
                        "COM1")
  --mapfile MAPFILE, -m MAPFILE
                        path to GeoJSON mapfile (e.g. "map.geojson")
  --htmlfile HTMLFILE, -o HTMLFILE
                        path to HTML output file (e.g. "map.html")
```
=>editing of GeoJSON file (editing of locations, modules and text):   
it's simple with website   
http://geojson.io   
simply "Open"->"File", then after editing "Save"->"GeoJSON"   
   
## example   
```
radioline_map.exe --port COM3 --mapfile ..\test\map.geojson --htmlfile ..\test\map.html --zoom 9 --open_browser -v -v
```   
   **=>subdirectory "test" on GitHub contains files for the example above.**   
   [hint: locations in "map.geojson" are fictional, the modules-setup is currently under my desk... :-) ]   
   [hint2: hmm, it seems that a HTML tag is missing in the generated file... otherwise I could you the interactive map under https://htmlpreview.github.io/?https://github.com/stefanbraun-private/radioline_map/blob/master/test/map.html ]   
   
Non-interactive Screenshot showing master module, network topology, RSSI-values, link quality and module metadata as popup and tooltip:   
![screenshot_radioline_map.png](https://stefanbraun-private.github.io/img/screenshot_radioline_map.png)
   
   
   
   
Disclaimer: Use 'radioline_map' at your own risk!   


CHANGES.txt
v0.0.1, January 13th 2019 -- release v0.0.1
v0.0.0, January 06th 2019 -- preparing start of project.

radioline_map/crc16_ifs.py
#!/usr/bin/env python
# encoding: utf-8
"""
radioline_serialtest/crc16_ifs.py

calculate CRC-16/IFS checksum.

(it seems that there exists many CRC implementations... :-/
but as "IO Ninja" showed and tested on https://crccalc.com/
RF modules *RAD-868-IFS* (c) by Phoenix Contact use the Modbus variant of CRC-16
but only using the lower byte of 16bit CRC value
)

Copyright (C) 2019 Stefan Braun



changelog:
v0.0.1, January 13th 2019 -- release v0.0.1
v0.0.0, January 03th 2019 -- preparing start of project.



This program is free software: you can redistribute it and/or modify it under the terms of the
GNU General Public License as published by the Free Software Foundation, either version 2 of the License,
or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.
If not, see <http://www.gnu.org/licenses/>.
"""


def calc_crc(data):
    ''' calculate CRC-16 in IFS-variant '''

    # special thanks to Kevin Herron
    # from https://stackoverflow.com/questions/39101926/port-modbus-rtu-crc-to-python-from-c-sharp
    crc = 0xFFFF
    for pos in data:
        crc ^= pos
        for i in range(8):
            if ((crc & 1) != 0):
                crc >>= 1
                crc ^= 0xA001
            else:
                crc >>= 1

    # difference to CRC16 as used in Modbus: only using lower byte
    return crc & 0xFF



if __name__ == '__main__':

    # test: telegram "80 42 02 C1" must have checksum "0x49"
    assert calc_crc(bytearray.fromhex("80 42 02 C1")) == 0x49

    # test: telegram "80 04 00 70 00 04 EE" must have checksum "0x03"
    assert calc_crc(bytearray.fromhex("80 04 00 70 00 04 EE")) == 0x03

    # test: telegram "80 04 00 10 00 08 EE" must have checksum "0x18"
    assert calc_crc(bytearray.fromhex("80 04 00 10 00 08 EE")) == 0x18

    # test: telegram "80 42 02 01 08 01 1C 00 00 00 00 51 05 EA ED 14 A9" must have checksum "0xA9"
    assert calc_crc(bytearray.fromhex("80 42 02 01 08 01 1C 00 00 00 00 51 05 EA ED 14")) == 0xA9

radioline_map/serialconn.py
#!/usr/bin/env python
# encoding: utf-8
"""
radioline_map/serialconn.py

*radioline_map* reads network topology from master RF-module *RAD-868-IFS* (c) by Phoenix Contact
 combines it with GeoJSON metadata and generates an interactive layer on OSM (Open Street Map).

Copyright (C) 2019 Stefan Braun


changelog:
v0.0.1, January 13th 2019 -- release v0.0.1
v0.0.0, January 06th 2019 -- preparing start of project.



This program is free software: you can redistribute it and/or modify it under the terms of the
GNU General Public License as published by the Free Software Foundation, either version 2 of the License,
or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.
If not, see <http://www.gnu.org/licenses/>.
"""

import logging
import serial
import array
from radioline_map.crc16_ifs import calc_crc



# setup of logging
# (based on tutorial https://docs.python.org/2/howto/logging.html )
# create logger =>set level to DEBUG if you want to catch all log messages!
logger = logging.getLogger('radioline_map.serialconn')
logger.setLevel(logging.DEBUG)



class Serialconn(object):
    """ serial connection to IFS-dataport of master RF-module """

    # timeout [s] for reading and writing
    SERIAL_TIMEOUT = 3.0

    # byte commands on IFS-dataport
    # with expected length of response
    # (reverse engineered: using "IO Ninja" for serial monitoring "PSI-CONF" talking to this RF modules)
    IFS_GET_RSSI = '80 04 2B 02 00 20 47 E7'
    IFS_GET_RSSI_LENGTH = 69
    IFS_GET_PARENTS = '80 04 2B 01 00 20 B7 E7'
    IFS_GET_PARENTS_LENGTH = 69


    STRING_PARENTS = 'parents'
    STRING_RSSI = 'rssi'

    def __init__(self, comport):
        self._comport = comport

    def _get_data(self, request_str, resp_length):
        """ open comport, send one request, check and return frame """

        # help from https://pyserial.readthedocs.io/en/latest/shortintro.html
        # and https://pyserial.readthedocs.io/en/latest/pyserial_api.html
        ser = serial.Serial()
        ser.port = self._comport
        ser.baudrate = 115200
        ser.bytesize = serial.EIGHTBITS
        ser.parity = serial.PARITY_EVEN
        ser.stopbits = serial.STOPBITS_ONE
        ser.timeout = Serialconn.SERIAL_TIMEOUT   # timeout for reading
        ser.xonxoff = False
        ser.rtscts = False
        ser.dsrdtr = False
        ser.write_timeout = 0

        logger.debug('Trying to open serialport "' + self._comport + '" with settings "115200,8,E,1,no flow control,timeout 5s"...')
        ser.open()

        # help from https://stackoverflow.com/questions/32018993/how-can-i-send-a-byte-array-to-a-serial-port-using-python
        data = bytearray.fromhex(request_str)
        logger.debug('Sending bytes "' + request_str + '" [total ' + str(len(data)) + ' bytes] to IFS-dataport...')
        ser.write(data)

        logger.debug('Trying to receive ' + str(resp_length) + ' bytes from IFS-dataport...')
        # read() blocks until number of bytes were received or timeout is over
        frame = ser.read(resp_length)
        if not frame:
            raise Exception('RF-module did not answer after ' + str(Serialconn.SERIAL_TIMEOUT) + ' seconds')

        # debugging: log response
        # help from https://stackoverflow.com/questions/6624453/whats-the-correct-way-to-convert-bytes-to-a-hex-string-in-python-3
        # FIXME: simple solution for printing hex numbers "AABBDD" in readable format "AA BB DD"...? :-/
        mylist = []
        for idx, char in enumerate(frame.hex()):
            mylist.append(char)
            if idx % 2 == 1:
                mylist.append(' ')
        frame_string = ''.join(mylist).strip()
        logger.debug('Received "' + frame_string + '" [total ' + str(len(frame)) + ' bytes]')

        # minimal check of received frame from attached "RAD-868-IFS"
        # frame length:
        assert len(frame) == resp_length, 'wrong frame length: expected ' + str(resp_length) + ' bytes, got ' + str(len(frame)) + ' bytes'

        # frame start:
        # help from https://stackoverflow.com/questions/16414559/how-to-use-hex-without-0x-in-python
        assert frame[0] == 0x80, 'wrong frame start character: expected "80", got "' + format(frame[0], 'x') + '"'

        # CRC over all bytes in frame:
        crc = calc_crc(frame[:-1])
        assert frame[-1] == crc, 'wrong checksum: expected "{}", got "{}"'.format(format(crc, 'x'), format(frame[-1], 'x'))

        logger.debug('Response from attached RF-module seems valid. :-)')

        # convert into array of unsigned integers
        # help from https://stackoverflow.com/questions/8461798/how-can-i-struct-unpack-many-numbers-at-once
        # and https://docs.python.org/3/library/array.html
        # (I had troubles with "struct", so I use "array" now...)
        arr = array.array('B')
        arr.frombytes(frame)
        return arr


    def get_networkstatus_list(self, subject_str):
        """ grab current network status from master module (parent-node and RSSI for every module) """

        if not subject_str in (Serialconn.STRING_PARENTS, Serialconn.STRING_RSSI):
            raise ValueError('"{}" or "{}" expected'.format(Serialconn.STRING_PARENTS, Serialconn.STRING_RSSI))

        logger.debug('Trying to retrieve ' + subject_str + ' for every slave module...')

        # =>we have to send four times the SAME request,
        #   the master RF module returns the data in parts
        if subject_str == Serialconn.STRING_PARENTS:
            request_str, resp_length = Serialconn.IFS_GET_PARENTS, Serialconn.IFS_GET_PARENTS_LENGTH
        else:
            request_str, resp_length = Serialconn.IFS_GET_RSSI, Serialconn.IFS_GET_RSSI_LENGTH

        parts_list = [None] * 4
        for x in range(4):
            arr = self._get_data(request_str=request_str,
                                 resp_length=resp_length)
            part_no = arr[3]
            parts_list[part_no] = arr[4:67]
            logger.debug('Retrieved ' + subject_str + ' part [' + str(part_no) + '/4]')

        # put together all parts for simpler handling
        # (fill index for invalid RAD-ID 0 and for master RAD-ID 1 with values same as for offline modules,
        # then index matches always the slave RAD-ID)
        if subject_str == Serialconn.STRING_PARENTS:
            offline_value = 0
        else:
            offline_value = 255
        networkstatus_list = [offline_value] * 2
        for part in parts_list:
            networkstatus_list.extend(part)

        # according to serial monitoring:
        # the raw array represents always RF modules 2 .. 250
        # =>depending on chosen RF module type and network mode not all indexes are used
        # =>the unnecessary trailing data looks always like offline modules (up to RAD-ID 250)
        # =>in our list with RAD-ID == list-index we have to cut three trailing zero-bytes
        return networkstatus_list[:-3]

radioline_map/generate_map.py
#!/usr/bin/env python
# encoding: utf-8
"""
radioline_map/generate_map.py

*radioline_map* reads network topology from master RF-module *RAD-868-IFS* (c) by Phoenix Contact
 combines it with GeoJSON metadata and generates an interactive layer on OSM (Open Street Map).

Copyright (C) 2019 Stefan Braun


changelog:
v0.0.1, January 13th 2019 -- release v0.0.1
v0.0.0, January 06th 2019 -- preparing start of project.



This program is free software: you can redistribute it and/or modify it under the terms of the
GNU General Public License as published by the Free Software Foundation, either version 2 of the License,
or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.
If not, see <http://www.gnu.org/licenses/>.
"""

import sys
import logging
import argparse
import json
import folium
from radioline_map.serialconn import Serialconn
import webbrowser
import os


# setup of logging
# (based on tutorial https://docs.python.org/2/howto/logging.html )
# create logger =>set level to DEBUG if you want to catch all log messages!
logger = logging.getLogger('radioline_map.generate_map')
logger.setLevel(logging.DEBUG)

# create console handler
# =>set level to DEBUG if you want to see everything on console!
ch = logging.StreamHandler()
ch.setLevel(logging.ERROR)

# create formatter
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')

# add formatter to ch
ch.setFormatter(formatter)

# add ch to logger
logger.addHandler(ch)

VERSION = '0.0.1'




class RadModule(object):
    def __init__(self, rad_id, postal_address, in_operation, longitude, latitude, comment):
        self.rad_id = rad_id
        self.postal_address = postal_address
        self.in_operation = in_operation
        self.longitude = longitude
        self.latitude = latitude
        self.comment = comment

        self.rssi = -255
        self.parent = 0

    def __repr__(self):
        """ developer representation of this object """
        # idea from https://stackoverflow.com/questions/25278563/python-return-dictionary-in-separate-lines-in-repr-method
        return self.__class__.__name__ + '(' + ', '.join('%s=%s' % (k, repr(v)) for k, v in self.__dict__.items()) + ')'



def main(comport, mapfile, htmlfile, zoom, open_browser, verbosity):

    logger.info('main(): radioline_map / version ' + VERSION)
    logger.info('main(): *********************************')

    # set console logging verbosity
    global ch
    if verbosity >= 3:
        ch.setLevel(logging.DEBUG)
    elif verbosity >= 2:
        ch.setLevel(logging.INFO)
    elif verbosity >= 1:
        ch.setLevel(logging.WARNING)
    else:
        ch.setLevel(logging.ERROR)


    logger.info('main(): Reading mapfile "' + mapfile + '" as GeoJSON file...')
    try:
        with open(mapfile, mode='r') as f:
            metadata = json.load(f)
    except Exception as ex:
        logger.exception('main(): Could not read mapfile! Got exception: ' + repr(ex))
        raise ex

    logger.info('main(): Extracting metadata from mapfile...')
    modules = {}
    try:
        rad_id_set = set()
        for feature in metadata['features']:
            # check RAD-ID
            rad_id = int(feature['properties']['RAD_ID'])
            assert 1 <= rad_id <= 250, 'RAD-ID must be a number from 1 to 250, got "' + str(rad_id) + '"instead...'
            assert rad_id not in rad_id_set, 'RAD-ID must be unique, "' + str(rad_id) + '" was chosen more than once...'
            rad_id_set.add(rad_id)

            # create Radioline module instances
            module = RadModule(rad_id=rad_id,
                               postal_address=feature['properties']['postal_address'],
                               in_operation=feature['properties']['in_operation'],
                               longitude=feature['geometry']['coordinates'][0],
                               latitude=feature['geometry']['coordinates'][1],
                               comment=feature['properties']['comment'])
            logger.debug('main(): Loaded Radioline-module: ' + repr(module))
            modules[rad_id] = module

        assert 1 in modules, 'Radioline master module with RAD-ID 1 is missing in GeoJSON file!'
    except Exception as ex:
        logger.exception('main(): Mapfile contains invalid metadata! Got exception: ' + repr(ex))
        raise ex


    # edited example from https://matplotlib.org/api/_as_gen/matplotlib.pyplot.plot.html
    #plt.plot([9.369685649871826, 8.723959922790527], [47.42320225367014, 47.500401963349994], 'r*-', linewidth=4)

    # generate layer on OpenStreetMap
    # based on examples from https://deparkes.co.uk/2016/05/06/colchester-public-toilets/
    # and https://deparkes.co.uk/2016/06/03/plot-lines-in-folium/

    # position of radioline master module is center of map
    logger.debug('main(): OpenStreetMap/leaflet.js map is centered at location [' + str(modules[1].longitude) + ', ' + str(
        modules[1].latitude) + ']')
    center = (modules[1].latitude, modules[1].longitude)
    rad_map = folium.Map(location=center, zoom_start=zoom)


    # get current network topology
    logger.info('main(): Establishing serial connection to master RF-module for getting network topology...')
    serialconn = Serialconn(comport=comport)
    parents_list = serialconn.get_networkstatus_list('parents')
    logger.debug('main(): parents_list=' + repr(parents_list))
    rssi_list = serialconn.get_networkstatus_list('rssi')
    logger.debug('main(): rssi_list=' + repr(rssi_list))


    # fill live network status into every slave module instance
    # =>we assume that GeoJSON file contains all RF modules!
    for key, mod in modules.items():
        # simplicity (or fragility?) in our data structure:
        #   key of modules dictionary == index in lists == RAD-ID
        mod.parent = parents_list[key]
        # interpretation of RSSI values: raw value 100 means -100dBm
        mod.rssi = -rssi_list[key]


    # add locations and connections to map
    for _, mod in modules.items():
        # adding module locations as markers
        if mod.rad_id == 1:
            # master module gets a special color
            color = 'blue'
        else:
            if mod.in_operation:
                if mod.parent:
                    color = 'green'
                else:
                    # unreachable module
                    color = 'red'
            else:
                # inactive module
                color = 'black'

        tooltip_str = '<br>'.join(['RAD-ID: ' + str(mod.rad_id),
                                   'RSSI: ' + str(mod.rssi) + 'dBm'])
        popup_str = '<br>'.join(['RAD-ID: ' + str(mod.rad_id),
                                 'address: ' + mod.postal_address,
                                 'RSSI: ' + str(mod.rssi) + 'dBm',
                                 'in operation: ' + 'yes' if mod.in_operation else 'no',
                                 'comment: ' + mod.comment])
        folium.Marker(location=[mod.latitude, mod.longitude],
                      popup=popup_str,
                      tooltip=tooltip_str,
                      icon=folium.Icon(color=color)).add_to(rad_map)

        # adding radio links as lines
        if mod.in_operation:
            if mod.parent:
                try:
                    parent = modules[mod.parent]
                    points = [(mod.latitude, mod.longitude),
                              (parent.latitude, parent.longitude)]
                    if mod.rssi > -90:
                        color = 'green'
                    elif mod.rssi > -100:
                        color = 'orange'
                    else:
                        color = 'red'
                    folium.PolyLine(points, color=color, weight=2.5, opacity=1).add_to(rad_map)
                except KeyError as ex:
                    logger.exception('main(): invalid topology detected: RAD-ID "' + str(mod.rad_id) + '" has invalid parent module!')
                    raise ex

    logger.info('main(): Saving generated OpenStreetMap/leaflet.js map into file ' + htmlfile)
    rad_map.save(htmlfile)

    if open_browser:
        # open generated map in system defaults webbrowser
        # with help from https://stackoverflow.com/questions/22004498/webbrowser-open-in-python
        abs_path = os.path.realpath(htmlfile)
        url = 'file://' + abs_path
        logger.info('main(): Open generated file in default webbrowser (URL=' + url + ')')
        webbrowser.open(url)

    # FIXME: we should return correct status when something went wrong...
    return 0        # success



if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Show network topology of RF-modules *RAD-868-IFS* on OSM(Open Street Map).')

    parser.add_argument('--zoom', '-z', dest='zoom', type=int, help='initial zoom factor on map (default: 15)', default=15)
    parser.add_argument('--open_browser', '-b', action="store_true", help='open generated map in webbrowser (default: False)')

    # allow increase of loglevel with command line argument
    # (help from https://docs.python.org/3/howto/argparse.html )
    parser.add_argument("-v", "--verbosity", action="count", default=0, help="output verbosity, setting this flag more times means more verbosity (default: loglevel WARNING)")

    # with help from https://stackoverflow.com/questions/24180527/argparse-required-arguments-listed-under-optional-arguments
    requiredNamed = parser.add_argument_group('required named arguments')

    requiredNamed.add_argument('--port', '-p', dest='comport', type=str, help='serial COM-port connected to IFS-dataport (e.g. "COM1")', required=True)
    requiredNamed.add_argument('--mapfile', '-m', dest='mapfile', type=str, help='path to GeoJSON mapfile (e.g. "map.geojson")', required=True)
    requiredNamed.add_argument('--htmlfile', '-o', dest='htmlfile', type=str, help='path to HTML output file (e.g. "map.html")', required=True)

    args = parser.parse_args()

    status = main(comport = args.comport,
                  mapfile = args.mapfile,
                  htmlfile = args.htmlfile,
                  zoom = args.zoom,
                  open_browser = args.open_browser,
                  verbosity = args.verbosity
                  )
    #sys.exit(status)
